In [2]:
#Data Loading

import pandas as pd

df = pd.read_csv("cs-training.csv")
df.head()

,Unnamed: 0,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30-59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60-89DaysPastDueNotWorse,NumberOfDependents
0,1,1,0.766127,45,2,0.802982,9120.0,13,0,6,0,2.0
1,2,0,0.957151,40,0,0.121876,2600.0,4,0,0,0,1.0
2,3,0,0.658180,38,1,0.085113,3042.0,2,1,0,0,0.0
3,4,0,0.233810,30,0,0.036050,3300.0,5,0,0,0,0.0
4,5,0,0.907239,49,1,0.024926,63588.0,7,0,1,0,0.0


In [3]:
#Data Exploration

df.info()
df.describe()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 12 columns):
 #   Column                                Non-Null Count   Dtype  
---  ------                                --------------   -----  
 0   Unnamed: 0                            150000 non-null  int64  
 1   SeriousDlqin2yrs                      150000 non-null  int64  
 2   RevolvingUtilizationOfUnsecuredLines  150000 non-null  float64
 3   age                                   150000 non-null  int64  
 4   NumberOfTime30-59DaysPastDueNotWorse  150000 non-null  int64  
 5   DebtRatio                             150000 non-null  float64
 6   MonthlyIncome                         120269 non-null  float64
 7   NumberOfOpenCreditLinesAndLoans       150000 non-null  int64  
 8   NumberOfTimes90DaysLate               150000 non-null  int64  
 9   NumberRealEstateLoansOrLines          150000 non-null  int64  
 10  NumberOfTime60-89DaysPastDueNotWorse  150000 non-null  int64  
 11  

Unnamed: 0                                  0
SeriousDlqin2yrs                            0
RevolvingUtilizationOfUnsecuredLines        0
age                                         0
NumberOfTime30-59DaysPastDueNotWorse        0
DebtRatio                                   0
MonthlyIncome                           29731
NumberOfOpenCreditLinesAndLoans             0
NumberOfTimes90DaysLate                     0
NumberRealEstateLoansOrLines                0
NumberOfTime60-89DaysPastDueNotWorse        0
NumberOfDependents                       3924
dtype: int64

In [4]:
# First Data Cleaning

# Eliminate Index Columns
df = df.drop(columns=['Unnamed: 0'], errors='ignore')

# Nuls
df.isnull().sum()

SeriousDlqin2yrs                            0
RevolvingUtilizationOfUnsecuredLines        0
age                                         0
NumberOfTime30-59DaysPastDueNotWorse        0
DebtRatio                                   0
MonthlyIncome                           29731
NumberOfOpenCreditLinesAndLoans             0
NumberOfTimes90DaysLate                     0
NumberRealEstateLoansOrLines                0
NumberOfTime60-89DaysPastDueNotWorse        0
NumberOfDependents                       3924
dtype: int64

In [5]:
# Handling Missing Values

# MonthlyIncome → median imputation (robust to outliers)
df['MonthlyIncome'].fillna(df['MonthlyIncome'].median(), inplace=True)

# NumberOfDependents → mode imputation (most common value)
df['NumberOfDependents'].fillna(df['NumberOfDependents'].mode()[0], inplace=True)

# Verify missing values
df.isnull().sum()

SeriousDlqin2yrs                        0
RevolvingUtilizationOfUnsecuredLines    0
age                                     0
NumberOfTime30-59DaysPastDueNotWorse    0
DebtRatio                               0
MonthlyIncome                           0
NumberOfOpenCreditLinesAndLoans         0
NumberOfTimes90DaysLate                 0
NumberRealEstateLoansOrLines            0
NumberOfTime60-89DaysPastDueNotWorse    0
NumberOfDependents                      0
dtype: int64

In [6]:
# Outlier Detection

# Summary statistics to identify extreme values
df.describe()

,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30-59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60-89DaysPastDueNotWorse,NumberOfDependents
count,150000.000000,150000.000000,150000.000000,150000.000000,150000.000000,1.500000e+05,150000.000000,150000.000000,150000.000000,150000.000000,150000.000000
mean,0.066840,6.048438,52.295207,0.421033,353.005076,6.418455e+03,8.452760,0.265973,1.018240,0.240387,0.737413
std,0.249746,249.755371,14.771866,4.192781,2037.818523,1.289040e+04,5.145951,4.169304,1.129771,4.155179,1.107021
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.029867,41.000000,0.000000,0.175074,3.903000e+03,5.000000,0.000000,0.000000,0.000000,0.000000
50%,0.000000,0.154181,52.000000,0.000000,0.366508,5.400000e+03,8.000000,0.000000,1.000000,0.000000,0.000000
75%,0.000000,0.559046,63.000000,0.000000,0.868254,7.400000e+03,11.000000,0.000000,2.000000,0.000000,1.000000
max,1.000000,50708.000000,109.000000,98.000000,329664.000000,3.008750e+06,58.000000,98.000000,54.000000,98.000000,20.000000


## Outlier Treatment

Extreme values were identified and removed based on business logic and statistical thresholds to ensure realistic data.

In [7]:
# Store original dataset shape before outlier treatment

original_shape = df.shape
original_shape

(150000, 11)

In [8]:
# Outlier Treatment

# Remove unrealistic ages
df = df[df['age'] > 18]

# Limit extreme DebtRatio values
df = df[df['DebtRatio'] < 5]

# Limit credit utilization to realistic range (<= 1)
df = df[df['RevolvingUtilizationOfUnsecuredLines'] <= 1]

# Remove extreme income outliers (top 1%)
df = df[df['MonthlyIncome'] < df['MonthlyIncome'].quantile(0.99)]

In [9]:
# Final Data Validation

cleaned_shape = df.shape

print("Before cleaning:", original_shape)
print("After cleaning:", cleaned_shape)

df.describe()

Before cleaning: (150000, 11)
After cleaning: (116182, 11)


,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30-59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60-89DaysPastDueNotWorse,NumberOfDependents
count,116182.000000,116182.000000,116182.000000,116182.000000,116182.000000,116182.000000,116182.00000,116182.000000,116182.000000,116182.000000,116182.000000
mean,0.062789,0.315471,51.673288,0.441945,0.368476,6272.608184,8.67717,0.281395,1.032389,0.260858,0.827168
std,0.242585,0.339008,14.654688,4.488780,0.411911,3826.980804,5.19125,4.465965,1.110142,4.455626,1.136100
min,0.000000,0.000000,21.000000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.031463,41.000000,0.000000,0.135860,3500.000000,5.00000,0.000000,0.000000,0.000000,0.000000
50%,0.000000,0.164066,51.000000,0.000000,0.289696,5400.000000,8.00000,0.000000,1.000000,0.000000,0.000000
75%,0.000000,0.544965,62.000000,0.000000,0.470183,8127.750000,11.00000,0.000000,2.000000,0.000000,1.000000
max,1.000000,1.000000,105.000000,98.000000,4.994445,24990.000000,57.00000,98.000000,25.000000,98.000000,20.000000


## Data Reduction

After applying outlier filtering, the dataset was reduced from 150,000 to 116,182 observations.

Approximately 22% of the data was removed due to unrealistic or extreme values. This step ensures higher data quality and more reliable analysis by focusing on realistic financial behavior.

## Feature Engineering

New features were created to better capture financial behavior and credit risk patterns.

In [10]:
# Feature Engineering

# Debt to income ratio (financial pressure indicator)
df['debt_to_income'] = df['DebtRatio']

# Income per dependent (household burden)
df['income_per_person'] = df['MonthlyIncome'] / (df['NumberOfDependents'] + 1)

# Total delinquency count
df['total_delinquency'] = (
    df['NumberOfTime30-59DaysPastDueNotWorse'] +
    df['NumberOfTimes90DaysLate'] +
    df['NumberOfTime60-89DaysPastDueNotWorse']
)

# Credit lines per person
df['credit_lines_per_person'] = df['NumberOfOpenCreditLinesAndLoans'] / (df['NumberOfDependents'] + 1)

These features help capture:

- Financial stress (debt vs income)
- Household burden
- Payment behavior
- Credit exposure

These are key drivers in real-world credit risk models.

In [11]:
df[['debt_to_income', 'income_per_person', 'total_delinquency']].describe()

,debt_to_income,income_per_person,total_delinquency
count,116182.000000,116182.000000,116182.000000
mean,0.368476,4320.275477,0.984197
std,0.411911,3227.451829,13.365941
min,0.000000,0.000000,0.000000
25%,0.135860,2016.666667,0.000000
50%,0.289696,3475.000000,0.000000
75%,0.470183,5556.000000,0.000000
max,4.994445,24990.000000,294.000000


## Feature Insights

The engineered features reveal important patterns:

- Most customers show low delinquency levels, but a small group presents extremely high values, indicating high-risk behavior.
- Debt-to-income ratio varies significantly, highlighting differences in financial stress across customers.
- Income distribution shows potential anomalies (e.g., zero income), which may require further validation.

These features provide strong signals for credit risk segmentation.

## Risk Scoring Model

A simple rule-based scoring system was developed to classify customers based on their credit risk.

In [12]:
# Initialize score
df['risk_score'] = 0

# High debt burden
df.loc[df['debt_to_income'] > 0.4, 'risk_score'] += 2

# Low income per person
df.loc[df['income_per_person'] < 2000, 'risk_score'] += 2

# Delinquency behavior
df.loc[df['total_delinquency'] > 0, 'risk_score'] += 3

# High credit utilization
df.loc[df['RevolvingUtilizationOfUnsecuredLines'] > 0.5, 'risk_score'] += 2

In [13]:
# Risk categories
def classify_risk(score):
    if score <= 2:
        return "Low Risk"
    elif score <= 5:
        return "Medium Risk"
    else:
        return "High Risk"

df['risk_category'] = df['risk_score'].apply(classify_risk)

In [14]:
df['risk_category'].value_counts()

risk_category
Low Risk       74972
Medium Risk    30228
High Risk      10982
Name: count, dtype: int64

## Risk Segmentation

The scoring model classified customers into three risk categories:

- 64% Low Risk
- 26% Medium Risk
- 9% High Risk

Most customers fall into the low-risk category, indicating a generally healthy portfolio. However, a smaller segment represents high-risk clients with significantly higher probability of default.

In [15]:
df.groupby('risk_category')[[
    'MonthlyIncome',
    'debt_to_income',
    'total_delinquency',
    'RevolvingUtilizationOfUnsecuredLines'
]].mean()

,MonthlyIncome,debt_to_income,total_delinquency,RevolvingUtilizationOfUnsecuredLines
risk_category,,,,
High Risk,4217.782826,0.663319,4.196321,0.742035
Low Risk,6819.013725,0.282363,0.000000,0.195067
Medium Risk,5663.933307,0.474938,2.258237,0.459125


In [16]:
df.groupby('risk_category')['SeriousDlqin2yrs'].mean()

risk_category
High Risk      0.237297
Low Risk       0.019754
Medium Risk    0.106127
Name: SeriousDlqin2yrs, dtype: float64

## Risk Validation

The relationship between risk categories and default rates confirms the effectiveness of the scoring model.

Higher risk categories show significantly higher default rates, validating the model's ability to differentiate customer risk profiles.

## Risk Profile Analysis

Clear differences emerge between risk segments:

- High Risk customers show lower income, higher debt levels, and significantly higher delinquency.
- Low Risk customers demonstrate strong financial stability, with higher income and no delinquency.
- Medium Risk customers fall in between, representing a potential opportunity segment.

These patterns confirm that financial behavior and repayment history are key drivers of credit risk.

## Business Recommendations

Based on the analysis, the following strategies are recommended:

- Limit loan approvals for high-risk customers or apply stricter conditions (e.g., lower loan amounts, higher interest rates)
- Prioritize low-risk customers for loan expansion strategies
- Develop targeted policies for medium-risk customers to maximize growth while managing risk
- Implement early warning systems based on delinquency behavior

These actions can help reduce default rates while maintaining portfolio growth.

In [17]:
# Export cleaned dataset
df.to_csv("credit_risk_cleaned.csv", index=False)

In [18]:
df.groupby('risk_category')['MonthlyIncome'].mean()

risk_category
High Risk      4217.782826
Low Risk       6819.013725
Medium Risk    5663.933307
Name: MonthlyIncome, dtype: float64

In [19]:
df.groupby('risk_category')['MonthlyIncome'].median()

risk_category
High Risk      3684.0
Low Risk       6000.0
Medium Risk    4974.0
Name: MonthlyIncome, dtype: float64

## Income Analysis by Risk Segment

Income shows a strong relationship with credit risk:

- Low-risk customers have significantly higher income levels compared to high-risk customers
- The median values confirm that this pattern is consistent and not driven by outliers
- Income distribution appears right-skewed, with a small number of high-income individuals

This suggests that income is a key factor in determining creditworthiness.

## Feature Importance Analysis

To better understand which variables are most predictive of default risk, a correlation analysis was performed.

In [20]:
# Correlation with target (default)
correlation = df.corr(numeric_only=True)['SeriousDlqin2yrs'].sort_values(ascending=False)

correlation

SeriousDlqin2yrs                        1.000000
risk_score                              0.289399
RevolvingUtilizationOfUnsecuredLines    0.241974
NumberOfTime30-59DaysPastDueNotWorse    0.121547
NumberOfTimes90DaysLate                 0.113184
total_delinquency                       0.112426
NumberOfTime60-89DaysPastDueNotWorse    0.101357
DebtRatio                               0.050329
debt_to_income                          0.050329
NumberOfDependents                      0.044155
NumberRealEstateLoansOrLines            0.004721
NumberOfOpenCreditLinesAndLoans        -0.021113
credit_lines_per_person                -0.036592
MonthlyIncome                          -0.056236
income_per_person                      -0.065142
age                                    -0.097526
Name: SeriousDlqin2yrs, dtype: float64

## Key Drivers of Default Risk

The analysis shows that delinquency-related variables have the strongest correlation with default risk.

This confirms that past payment behavior is a more reliable predictor than demographic or income-related variables.

Credit utilization and debt levels also contribute to risk, but to a lesser extent.

## Data Extraction (SQL)

The data used in this analysis was extracted from a relational database using SQL.

The following query was used to combine customer, loan, and payment data: